delegate-app/Main.hs

```diff
@@ -1,62 +1,254 @@
+    writeTQueue q (DelegateEvent Start)
+    pure q
+  delegateBroadCastChan <- liftIO . atomically $ newTChan
+
+  let wsServer, delegateRunner, queueAuctionPhases :: IO ()
+      wsServer =
+        runServer (show $ host conf) (fromIntegral $ port conf) $
+          runWithTracer' tracer . delegateServerApp (ping conf) delegateInputChan delegateBroadCastChan
+
+      delegateRunner =
+        runWithTracer' tracer . execDelegateRunnerT $ mkRunner (tick conf) delegateInputChan delegateBroadCastChan
+
+      queueAuctionPhases = runWithTracer' tracer $ mbQueueAuctionPhases (tick conf) delegateInputChan delegateBroadCastChan
 
-consumer :: Chan DelegateInput -> DelegateRunnerT IO ()
-consumer delegateInputs = do
-  -- This will block until an event appears in the queue
-  delegateIn <- liftIO $ readChan delegateInputs
-  delegateResponse <- delegateStep delegateIn
-  liftIO $ mbQueueAuctionPhases delegateResponse delegateInputs
-  -- FIXME: send response to client
   liftIO $
-    putStrLn $
-      "Delegate responses for input: " <> show delegateResponse
+    mapConcurrently_
+      (runWithTracer' tracer . uncurry startWorker)
+      [ (WebsocketThread, wsServer)
+      , (QueueAuctionStageThread, queueAuctionPhases)
+      , (DelegateRunnerThread, delegateRunner)
+      ]
+
+-- | start a worker and log it
+startWorker :: ThreadSort -> IO () -> DelegateTracerT IO ()
+startWorker thread act = do
+  trace $ ThreadEvent ThreadStarted thread
+  liftIO act
+  trace $ ThreadEvent ThreadCancelled thread
 
-mbQueueAuctionPhases :: [DelegateResponse] -> Chan DelegateInput -> IO ()
-mbQueueAuctionPhases [AuctionSet terms] events = void $ async queueCurrentStage
+mbQueueAuctionPhases :: Int -> TQueue DelegateInput -> TChan DelegateResponse -> DelegateTracerT IO ()
+mbQueueAuctionPhases tick reqq broadcast = do
+  terms <- liftIO . atomically $ getTerms =<< dupTChan broadcast
+  liftIO $ queueStages terms
+  traceQueueEvent terms
   where
-    queueCurrentStage = do
+    traceQueueEvent :: AuctionTerms -> DelegateTracerT IO ()
+    traceQueueEvent = trace . QueueAuctionPhaseEvent . ReceivedAuctionSet
+
+    getTerms :: TChan DelegateResponse -> STM AuctionTerms
+    getTerms chan =
+      readTChan chan >>= \case
+        AuctionSet terms -> pure terms
+        _ -> getTerms chan
+
+    queueStages :: AuctionTerms -> IO ()
+    queueStages terms = do
       currentStage <- currentAuctionStage terms
-      writeChan events (DelegateEvent $ AuctionStageStarted currentStage)
+      atomically $ writeTQueue reqq (DelegateEvent $ AuctionStageStarted currentStage)
       currentPosixTime <- POSIXTime <$> currentTimeMilliseconds
       let mSecsLeft = secondsLeftInInterval currentPosixTime (stageToInterval terms currentStage)
       case mSecsLeft of
         Nothing -> pure ()
         Just s -> do
-          threadDelay (fromInteger s * 1000)
-          queueCurrentStage
-mbQueueAuctionPhases _ _ = pure ()
+          threadDelay (fromInteger s * tick)
```

Comment:
I don't think this is right... The 1000 is meant to convert seconds to milliseconds, so threadDelay waits the correct amount of time

Reply:
Oh. Thank you. Will fix